// incur@0.1.17/dist/Cli.js
import * as Completions from './Completions.js';
import { IncurError, ValidationError } from './Errors.js';
import * as Formatter from './Formatter.js';
import * as Help from './Help.js';
import { detectRunner } from './internal/pm.js';
import * as Mcp from './Mcp.js';
import * as Parser from './Parser.js';
import * as Schema from './Schema.js';
import * as Skill from './Skill.js';
import * as SyncMcp from './SyncMcp.js';
import * as SyncSkills from './SyncSkills.js';
export function create(nameOrDefinition, definition) {
    const name = typeof nameOrDefinition === 'string' ? nameOrDefinition : nameOrDefinition.name;
    const def = typeof nameOrDefinition === 'string' ? (definition ?? {}) : nameOrDefinition;
    const rootDef = 'run' in def ? def : undefined;
    const commands = new Map();
    const middlewares = [];
    const cli = {
        name,
        description: def.description,
        env: def.env,
        vars: def.vars,
        command(nameOrCli, def) {
            if (typeof nameOrCli === 'string') {
                commands.set(nameOrCli, def);
                return cli;
            }
            const mountedRootDef = toRootDefinition.get(nameOrCli);
            if (mountedRootDef) {
                commands.set(nameOrCli.name, mountedRootDef);
                return cli;
            }
            const sub = nameOrCli;
            const subCommands = toCommands.get(sub);
            const subOutputPolicy = toOutputPolicy.get(sub);
            const subMiddlewares = toMiddlewares.get(sub);
            commands.set(sub.name, {
                _group: true,
                description: sub.description,
                commands: subCommands,
                ...(subOutputPolicy ? { outputPolicy: subOutputPolicy } : undefined),
                ...(subMiddlewares?.length ? { middlewares: subMiddlewares } : undefined),
            });
            return cli;
        },
        async serve(argv = process.argv.slice(2), serveOptions = {}) {
            return serveImpl(name, commands, argv, {
                ...serveOptions,
                aliases: def.aliases,
                description: def.description,
                envSchema: def.env,
                format: def.format,
                mcp: def.mcp,
                middlewares,
                outputPolicy: def.outputPolicy,
                rootCommand: rootDef,
                sync: def.sync,
                vars: def.vars,
                version: def.version,
            });
        },
        use(handler) {
            middlewares.push(handler);
            return cli;
        },
    };
    if (rootDef)
        toRootDefinition.set(cli, rootDef);
    if (def.outputPolicy)
        toOutputPolicy.set(cli, def.outputPolicy);
    toMiddlewares.set(cli, middlewares);
    toCommands.set(cli, commands);
    return cli;
}
/** @internal Shared serve implementation for both router and leaf CLIs. */
// biome-ignore lint/correctness/noUnusedVariables: _
async function serveImpl(name, commands, argv, options = {}) {
    const stdout = options.stdout ?? ((s) => process.stdout.write(s));
    const exit = options.exit ?? ((code) => process.exit(code));
    const { verbose, format: formatFlag, formatExplicit, llms, mcp: mcpFlag, help, version, rest: filtered, } = extractBuiltinFlags(argv);
    // --mcp: start as MCP stdio server
    if (mcpFlag) {
        await Mcp.serve(name, options.version ?? '0.0.0', commands);
        return;
    }
    // COMPLETE: dynamic shell completions (called by shell hook at tab-press)
    const completeShell = process.env.COMPLETE;
    if (completeShell) {
        // Remove separator `--` from argv
        const sepIdx = argv.indexOf('--');
        const words = sepIdx !== -1 ? argv.slice(sepIdx + 1) : argv;
        if (words.length === 0) {
            // Registration mode: print shell hook script for primary name + aliases
            const names = [name, ...(options.aliases ?? [])];
            stdout(names.map((n) => Completions.register(completeShell, n)).join('\n'));
        }
        else {
            const index = Number(process.env._COMPLETE_INDEX ?? words.length - 1);
            const candidates = Completions.complete(commands, options.rootCommand, words, index);
            const out = Completions.format(completeShell, candidates);
            if (out)
                stdout(out);
        }
        return;
    }
    // Human mode: stdout is a TTY.
    const human = process.stdout.isTTY === true;
    function writeln(s) {
        stdout(s.endsWith('\n') ? s : `${s}\n`);
    }
    // Skills staleness check (skip for built-in commands)
    if (!llms && !help && !version) {
        const isSkillsAdd = filtered[0] === 'skills' || (filtered[0] === name && filtered[1] === 'skills');
        const isMcpAdd = filtered[0] === 'mcp' || (filtered[0] === name && filtered[1] === 'mcp');
        if (!isSkillsAdd && !isMcpAdd) {
            const stored = SyncSkills.readHash(name);
            if (stored) {
                const groups = new Map();
                const entries = collectSkillCommands(commands, [], groups);
                if (Skill.hash(entries) !== stored) {
                    const runner = detectRunner();
                    const spec = SyncMcp.detectPackageSpecifier(name);
                    process.stderr.write(`⚠ Skills are out of date. Run '${runner} ${spec} skills add' to update.\n\n`);
                }
            }
        }
    }
    if (llms) {
        // Scope to a subtree if command tokens are provided
        let scopedCommands = commands;
        const prefix = [];
        for (const token of filtered) {
            const entry = scopedCommands.get(token);
            if (!entry)
                break;
            if (isGroup(entry)) {
                scopedCommands = entry.commands;
                prefix.push(token);
            }
            else {
                // Leaf command — scope to just this command
                scopedCommands = new Map([[token, entry]]);
                break;
            }
        }
        if (!formatExplicit || formatFlag === 'md') {
            const groups = new Map();
            const cmds = collectSkillCommands(scopedCommands, prefix, groups);
            const scopedName = prefix.length > 0 ? `${name} ${prefix.join(' ')}` : name;
            writeln(Skill.generate(scopedName, cmds, groups));
            return;
        }
        writeln(Formatter.format(buildManifest(scopedCommands, prefix), formatFlag));
        return;
    }
    // completions <shell>: print shell hook script to stdout
    const completionsIdx = (() => {
        // e.g. `completions bash`
        if (filtered[0] === 'completions')
            return 0;
        // e.g. `my-cli completions bash`
        if (filtered[0] === name && filtered[1] === 'completions')
            return 1;
        // not a completions invocation
        return -1;
    })();
    if (completionsIdx !== -1 && filtered[completionsIdx] === 'completions') {
        if (help) {
            writeln([
                `${name} completions — Generate shell completion script`,
                '',
                `Usage: ${name} completions <shell>`,
                '',
                'Shells:',
                '  bash',
                '  fish',
                '  nushell',
                '  zsh',
                '',
                'Setup:',
                ...(() => {
                    const rows = [
                        ['bash', `eval "$(${name} completions bash)"`, '# add to ~/.bashrc'],
                        ['zsh', `eval "$(${name} completions zsh)"`, '# add to ~/.zshrc'],
                        ['fish', `${name} completions fish | source`, '# add to ~/.config/fish/config.fish'],
                        ['nushell', `see \`${name} completions nushell\``, '# add to config.nu'],
                    ];
                    const shellW = Math.max(...rows.map((r) => r[0].length));
                    const cmdW = Math.max(...rows.map((r) => r[1].length));
                    return rows.map(([shell, cmd, comment]) => `  ${shell.padEnd(shellW)}  ${cmd.padEnd(cmdW)}  ${comment}`);
                })(),
            ].join('\n'));
            return;
        }
        const shell = filtered[completionsIdx + 1];
        if (!shell || !['bash', 'fish', 'nushell', 'zsh'].includes(shell)) {
            writeln(formatHumanError({
                code: 'INVALID_SHELL',
                message: shell
                    ? `Unknown shell '${shell}'. Supported: bash, fish, nushell, zsh`
                    : `Missing shell argument. Usage: ${name} completions <bash|fish|nushell|zsh>`,
            }));
            exit(1);
            return;
        }
        const names = [name, ...(options.aliases ?? [])];
        writeln(names.map((n) => Completions.register(shell, n)).join('\n'));
        return;
    }
    // skills add: generate skill files and install via `<pm>x skills add` (only when sync is configured)
    const skillsIdx = filtered[0] === 'skills' ? 0 : filtered[0] === name && filtered[1] === 'skills' ? 1 : -1;
    if (skillsIdx !== -1 && filtered[skillsIdx] === 'skills' && filtered[skillsIdx + 1] === 'add') {
        if (help) {
            writeln([
                `${name} skills add — Sync skill files to your agent`,
                '',
                `Usage: ${name} skills add [options]`,
                '',
                'Options:',
                '  --depth <number>  Grouping depth for skill files (default: 1)',
                '  --no-global       Install to project instead of globally',
            ].join('\n'));
            return;
        }
        const rest = filtered.slice(skillsIdx + 2);
        const depthArg = rest.indexOf('--depth');
        const depthEq = rest.find((t) => t.startsWith('--depth='));
        const depth = depthArg !== -1
            ? Number(rest[depthArg + 1])
            : depthEq
                ? Number(depthEq.split('=')[1])
                : (options.sync?.depth ?? 1);
        const global = rest.includes('--no-global') ? false : undefined;
        try {
            stdout('Syncing...');
            const result = await SyncSkills.sync(name, commands, {
                cwd: options.sync?.cwd,
                depth,
                description: options.description,
                global,
                include: options.sync?.include,
            });
            stdout('\r\x1b[K');
            const lines = [];
            const skillLabel = (s) => s.external || s.name === name ? s.name : `${name}-${s.name}`;
            const maxLen = Math.max(...result.skills.map((s) => skillLabel(s).length));
            for (const s of result.skills) {
                const label = skillLabel(s);
                const padding = s.description
                    ? `${' '.repeat(maxLen - label.length)}  ${s.description}`
                    : '';
                lines.push(`  ✓ ${label}${padding}`);
            }
            lines.push('');
            lines.push(`${result.skills.length} skill${result.skills.length === 1 ? '' : 's'} synced`);
            const suggestions = options.sync?.suggestions;
            if (suggestions && suggestions.length > 0) {
                lines.push('');
                lines.push(`Your agent can now use ${name}. Try asking:`);
                for (const s of suggestions)
                    lines.push(`  "${s}"`);
            }
            lines.push('');
            lines.push(`Run \`${name} --help\` to see the full command reference.`);
            writeln(lines.join('\n'));
            if (verbose || formatExplicit) {
                const output = { skills: result.paths };
                if (verbose && result.agents.length > 0)
                    output.agents = result.agents;
                writeln(Formatter.format(output, formatExplicit ? formatFlag : 'toon'));
            }
        }
        catch (err) {
            writeln(Formatter.format({ code: 'SYNC_SKILLS_FAILED', message: err instanceof Error ? err.message : String(err) }, formatExplicit ? formatFlag : 'toon'));
            exit(1);
        }
        return;
    }
    // mcp add: register CLI as MCP server via `npx add-mcp`
    const mcpIdx = filtered[0] === 'mcp' ? 0 : filtered[0] === name && filtered[1] === 'mcp' ? 1 : -1;
    if (mcpIdx !== -1 && filtered[mcpIdx] === 'mcp' && filtered[mcpIdx + 1] === 'add') {
        if (help) {
            writeln([
                `${name} mcp add — Register as an MCP server for your agent`,
                '',
                `Usage: ${name} mcp add [options]`,
                '',
                'Options:',
                '  -c, --command <cmd>  Override the command agents will run (e.g. "pnpm my-cli --mcp")',
                '  --no-global          Install to project instead of globally',
                '  --agent <agent>      Target a specific agent (e.g. claude-code, cursor)',
            ].join('\n'));
            return;
        }
        const rest = filtered.slice(mcpIdx + 2);
        const global = rest.includes('--no-global') ? false : true;
        // Parse --command / -c and --agent flags from argv
        let command = options.mcp?.command;
        const agents = [...(options.mcp?.agents ?? [])];
        for (let i = 0; i < rest.length; i++) {
            if ((rest[i] === '--command' || rest[i] === '-c') && rest[i + 1])
                command = rest[++i];
            else if (rest[i] === '--agent' && rest[i + 1])
                agents.push(rest[++i]);
        }
        try {
            stdout('Registering MCP server...');
            const result = await SyncMcp.register(name, {
                command,
                global,
                agents,
            });
            stdout('\r\x1b[K');
            const lines = [];
            lines.push(`✓ Registered ${name} as MCP server`);
            if (result.agents.length > 0)
                lines.push(`  Agents: ${result.agents.join(', ')}`);
            lines.push('');
            lines.push(`Agents can now use ${name} tools.`);
            const suggestions = options.sync?.suggestions;
            if (suggestions && suggestions.length > 0) {
                lines.push('');
                lines.push('Try asking:');
                for (const s of suggestions)
                    lines.push(`  "${s}"`);
            }
            writeln(lines.join('\n'));
            if (verbose || formatExplicit)
                writeln(Formatter.format({ name, command: result.command, agents: result.agents }, formatExplicit ? formatFlag : 'toon'));
        }
        catch (err) {
            writeln(Formatter.format({ code: 'MCP_ADD_FAILED', message: err instanceof Error ? err.message : String(err) }, formatExplicit ? formatFlag : 'toon'));
            exit(1);
        }
        return;
    }
    // --help takes precedence over --version
    if (version && !help && options.version) {
        writeln(options.version);
        return;
    }
    if (filtered.length === 0) {
        if (options.rootCommand &&
            human &&
            options.rootCommand.args &&
            hasRequiredArgs(options.rootCommand.args)) {
            // Root command with args but none provided (human mode) — show help
            const cmd = options.rootCommand;
            writeln(Help.formatCommand(name, {
                alias: cmd.alias,
                aliases: options.aliases,
                description: cmd.description ?? options.description,
                version: options.version,
                args: cmd.args,
                env: cmd.env,
                envSource: options.env,
                hint: cmd.hint,
                options: cmd.options,
                examples: formatExamples(cmd.examples),
                usage: cmd.usage,
                commands: commands.size > 0 ? collectHelpCommands(commands) : undefined,
                root: true,
            }));
            return;
        }
        if (options.rootCommand) {
            // Root command with no args — treat as root invocation
        }
        else {
            writeln(Help.formatRoot(name, {
                aliases: options.aliases,
                description: options.description,
                version: options.version,
                commands: collectHelpCommands(commands),
                root: true,
            }));
            return;
        }
    }
    const resolved = filtered.length === 0 && options.rootCommand
        ? { command: options.rootCommand, path: name, rest: [] }
        : resolveCommand(commands, filtered);
    // --help after a command → show help for that command
    if (help) {
        if ('help' in resolved || 'error' in resolved) {
            // group or unknown → show root help for that path
            const helpName = 'help' in resolved ? `${name} ${resolved.path}` : name;
            const helpDesc = 'help' in resolved ? resolved.description : options.description;
            const helpCmds = 'help' in resolved ? resolved.commands : commands;
            const isRoot = helpName === name;
            // Root with both a handler and subcommands → show command help with subcommands
            if (isRoot && options.rootCommand && helpCmds.size > 0) {
                const cmd = options.rootCommand;
                writeln(Help.formatCommand(name, {
                    alias: cmd.alias,
                    aliases: options.aliases,
                    description: cmd.description ?? options.description,
                    version: options.version,
                    args: cmd.args,
                    env: cmd.env,
                    envSource: options.env,
                    hint: cmd.hint,
                    options: cmd.options,
                    examples: formatExamples(cmd.examples),
                    usage: cmd.usage,
                    commands: collectHelpCommands(helpCmds),
                    root: true,
                }));
            }
            else {
                writeln(Help.formatRoot(helpName, {
                    aliases: isRoot ? options.aliases : undefined,
                    description: helpDesc,
                    version: isRoot ? options.version : undefined,
                    commands: collectHelpCommands(helpCmds),
                    root: isRoot,
                }));
            }
        }
        else {
            const isRootCmd = resolved.path === name;
            const commandName = isRootCmd ? name : `${name} ${resolved.path}`;
            const helpSubcommands = isRootCmd && options.rootCommand && commands.size > 0
                ? collectHelpCommands(commands)
                : undefined;
            writeln(Help.formatCommand(commandName, {
                alias: resolved.command.alias,
                aliases: isRootCmd ? options.aliases : undefined,
                description: resolved.command.description,
                version: isRootCmd ? options.version : undefined,
                args: resolved.command.args,
                env: resolved.command.env,
                envSource: options.env,
                hint: resolved.command.hint,
                options: resolved.command.options,
                examples: formatExamples(resolved.command.examples),
                usage: resolved.command.usage,
                commands: helpSubcommands,
                root: isRootCmd,
            }));
        }
        return;
    }
    if ('help' in resolved) {
        writeln(Help.formatRoot(`${name} ${resolved.path}`, {
            description: resolved.description,
            commands: collectHelpCommands(resolved.commands),
        }));
        return;
    }
    const start = performance.now();
    // Resolve effective format: explicit --format/--json → command default → CLI default → toon
    const resolvedFormat = 'command' in resolved && resolved.command.format;
    const format = formatExplicit ? formatFlag : resolvedFormat || options.format || 'toon';
    // Fall back to root command when no subcommand matches
    const effective = 'error' in resolved && options.rootCommand && !resolved.path
        ? { command: options.rootCommand, path: name, rest: filtered }
        : resolved;
    // Resolve outputPolicy: command/group → CLI-level → default ('all')
    const effectiveOutputPolicy = ('outputPolicy' in resolved && resolved.outputPolicy) || options.outputPolicy;
    const renderOutput = !(human && !formatExplicit && effectiveOutputPolicy === 'agent-only');
    function write(output) {
        const cta = output.meta.cta;
        if (human && !verbose) {
            if (output.ok && output.data != null && renderOutput)
                writeln(Formatter.format(output.data, format));
            else if (!output.ok)
                writeln(formatHumanError(output.error));
            if (cta)
                writeln(formatHumanCta(cta));
            return;
        }
        if (verbose)
            return writeln(Formatter.format(output, format));
        const base = output.ok ? output.data : output.error;
        const formatted = Formatter.format(base, format);
        if (!cta) {
            if (formatted)
                writeln(formatted);
            return;
        }
        const payload = typeof base === 'object' && base !== null ? { ...base, cta } : { data: base, cta };
        writeln(Formatter.format(payload, format));
    }
    if ('error' in effective) {
        const helpCmd = effective.path ? `${name} ${effective.path} --help` : `${name} --help`;
        const parent = effective.path ? `${name} ${effective.path}` : name;
        const message = `'${effective.error}' is not a command for '${parent}'.`;
        const cta = {
            description: 'See available commands:',
            commands: [{ command: helpCmd }],
        };
        if (human && !verbose) {
            writeln(formatHumanError({ code: 'COMMAND_NOT_FOUND', message }));
            writeln(formatHumanCta(cta));
            exit(1);
            return;
        }
        write({
            ok: false,
            error: { code: 'COMMAND_NOT_FOUND', message },
            meta: {
                command: effective.error,
                cta,
                duration: `${Math.round(performance.now() - start)}ms`,
            },
        });
        exit(1);
        return;
    }
    const { command, path, rest } = effective;
    // Collect middleware: root CLI + groups traversed + per-command
    const allMiddleware = [
        ...(options.middlewares ?? []),
        ...('middlewares' in resolved
            ? (resolved.middlewares ?? [])
            : []),
        ...(command.middleware ?? []),
    ];
    // Initialize vars from schema defaults
    const varsMap = options.vars ? options.vars.parse({}) : {};
    const envSource = options.env ?? process.env;
    const runCommand = async () => {
        const { args, options: parsedOptions } = Parser.parse(rest, {
            alias: command.alias,
            args: command.args,
            options: command.options,
        });
        if (human)
            emitDeprecationWarnings(rest, command.options, command.alias);
        const env = command.env ? Parser.parseEnv(command.env, envSource) : {};
        const okFn = (data, meta = {}) => {
            return { [sentinel]: 'ok', data, cta: meta.cta };
        };
        const errorFn = (opts) => {
            return { [sentinel]: 'error', ...opts };
        };
        const result = command.run({
            agent: !human,
            args,
            env,
            name,
            options: parsedOptions,
            ok: okFn,
            error: errorFn,
            var: varsMap,
            version: options.version,
        });
        // Streaming path — async generator
        if (isAsyncGenerator(result)) {
            await handleStreaming(result, {
                name,
                path,
                start,
                format,
                formatExplicit,
                human,
                renderOutput,
                verbose,
                write,
                writeln,
                exit,
            });
            return;
        }
        const awaited = await result;
        if (isSentinel(awaited)) {
            const cta = formatCtaBlock(name, awaited.cta);
            if (awaited[sentinel] === 'ok') {
                write({
                    ok: true,
                    data: awaited.data,
                    meta: {
                        command: path,
                        duration: `${Math.round(performance.now() - start)}ms`,
                        ...(cta ? { cta } : undefined),
                    },
                });
            }
            else {
                write({
                    ok: false,
                    error: {
                        code: awaited.code,
                        message: awaited.message,
                        ...(awaited.retryable !== undefined ? { retryable: awaited.retryable } : undefined),
                    },
                    meta: {
                        command: path,
                        duration: `${Math.round(performance.now() - start)}ms`,
                        ...(cta ? { cta } : undefined),
                    },
                });
                exit(1);
            }
        }
        else {
            write({
                ok: true,
                data: awaited,
                meta: {
                    command: path,
                    duration: `${Math.round(performance.now() - start)}ms`,
                },
            });
        }
    };
    try {
        const cliEnv = options.envSchema ? Parser.parseEnv(options.envSchema, envSource) : {};
        if (allMiddleware.length > 0) {
            const errorFn = (opts) => {
                return { [sentinel]: 'error', ...opts };
            };
            const mwCtx = {
                agent: !human,
                command: path,
                env: cliEnv,
                error: errorFn,
                name,
                set(key, value) {
                    varsMap[key] = value;
                },
                var: varsMap,
                version: options.version,
            };
            const handleMwSentinel = (result) => {
                if (!isSentinel(result) || result[sentinel] !== 'error')
                    return;
                const cta = formatCtaBlock(name, result.cta);
                write({
                    ok: false,
                    error: {
                        code: result.code,
                        message: result.message,
                        ...(result.retryable !== undefined ? { retryable: result.retryable } : undefined),
                    },
                    meta: {
                        command: path,
                        duration: `${Math.round(performance.now() - start)}ms`,
                        ...(cta ? { cta } : undefined),
                    },
                });
                exit(1);
            };
            const composed = allMiddleware.reduceRight((next, mw) => async () => {
                handleMwSentinel(await mw(mwCtx, next));
            }, runCommand);
            await composed();
        }
        else {
            await runCommand();
        }
    }
    catch (error) {
        const errorOutput = {
            ok: false,
            error: {
                code: error instanceof IncurError
                    ? error.code
                    : error instanceof ValidationError
                        ? 'VALIDATION_ERROR'
                        : 'UNKNOWN',
                message: error instanceof Error ? error.message : String(error),
                ...(error instanceof IncurError ? { retryable: error.retryable } : undefined),
                ...(error instanceof ValidationError ? { fieldErrors: error.fieldErrors } : undefined),
            },
            meta: {
                command: path,
                duration: `${Math.round(performance.now() - start)}ms`,
            },
        };
        if (human && !formatExplicit && error instanceof ValidationError) {
            writeln(formatHumanValidationError(name, path, command, error, options.env));
            exit(1);
            return;
        }
        write(errorOutput);
        exit(1);
    }
}
/** @internal Formats a validation error for TTY with usage hint. */
function formatHumanValidationError(cli, path, command, error, envSource) {
    const lines = [];
    for (const fe of error.fieldErrors)
        lines.push(`Error: missing required argument <${fe.path}>`);
    lines.push('See below for usage.');
    lines.push('');
    lines.push(Help.formatCommand(path === cli ? cli : `${cli} ${path}`, {
        alias: command.alias,
        description: command.description,
        args: command.args,
        env: command.env,
        envSource,
        hint: command.hint,
        options: command.options,
        examples: formatExamples(command.examples),
        usage: command.usage,
    }));
    return lines.join('\n');
}
/** @internal Resolves a command from the tree by walking tokens until a leaf is found. */
function resolveCommand(commands, tokens) {
    const [first, ...rest] = tokens;
    if (!first || !commands.has(first))
        return { error: first ?? '(none)', path: '' };
    let entry = commands.get(first);
    const path = [first];
    let remaining = rest;
    let inheritedOutputPolicy;
    const collectedMiddlewares = [];
    while (isGroup(entry)) {
        if (entry.outputPolicy)
            inheritedOutputPolicy = entry.outputPolicy;
        if (entry.middlewares)
            collectedMiddlewares.push(...entry.middlewares);
        const next = remaining[0];
        if (!next)
            return {
                help: true,
                path: path.join(' '),
                description: entry.description,
                commands: entry.commands,
            };
        const child = entry.commands.get(next);
        if (!child) {
            return { error: next, path: path.join(' ') };
        }
        path.push(next);
        remaining = remaining.slice(1);
        entry = child;
    }
    const outputPolicy = entry.outputPolicy ?? inheritedOutputPolicy;
    return {
        command: entry,
        middlewares: collectedMiddlewares,
        path: path.join(' '),
        rest: remaining,
        ...(outputPolicy ? { outputPolicy } : undefined),
    };
}
/** @internal Extracts built-in flags (--verbose, --format, --json, --llms, --help, --version) from argv. */
function extractBuiltinFlags(argv) {
    let verbose = false;
    let llms = false;
    let mcp = false;
    let help = false;
    let version = false;
    let format = 'toon';
    let formatExplicit = false;
    const rest = [];
    for (let i = 0; i < argv.length; i++) {
        const token = argv[i];
        if (token === '--verbose')
            verbose = true;
        else if (token === '--llms')
            llms = true;
        else if (token === '--mcp')
            mcp = true;
        else if (token === '--help' || token === '-h')
            help = true;
        else if (token === '--version')
            version = true;
        else if (token === '--json') {
            format = 'json';
            formatExplicit = true;
        }
        else if (token === '--format' && argv[i + 1]) {
            format = argv[i + 1];
            formatExplicit = true;
            i++;
        }
        else
            rest.push(token);
    }
    return { verbose, format, formatExplicit, llms, mcp, help, version, rest };
}
/** @internal Collects immediate child commands/groups for help output. */
function collectHelpCommands(commands) {
    const result = [];
    for (const [name, entry] of commands) {
        if (isGroup(entry))
            result.push({ name, description: entry.description });
        else
            result.push({ name, description: entry.description });
    }
    return result.sort((a, b) => a.name.localeCompare(b.name));
}
/** @internal Type guard for command groups. */
function isGroup(entry) {
    return '_group' in entry;
}
/** @internal Maps CLI instances to their command maps. */
export const toCommands = new WeakMap();
/** @internal Maps CLI instances to their middleware arrays. */
const toMiddlewares = new WeakMap();
/** @internal Maps root CLI instances to their command definitions. */
const toRootDefinition = new WeakMap();
/** @internal Maps CLI instances to their output policy. */
const toOutputPolicy = new WeakMap();
/** @internal Sentinel symbol for `ok()` and `error()` return values. */
const sentinel = Symbol.for('incur.sentinel');
/** @internal Formats an error for human-readable TTY output. */
function formatHumanError(error) {
    const prefix = error.code === 'UNKNOWN' || error.code === 'COMMAND_NOT_FOUND'
        ? 'Error'
        : `Error (${error.code})`;
    let out = `${prefix}: ${error.message}`;
    if (error.fieldErrors)
        for (const fe of error.fieldErrors)
            out += `\n  ${fe.path}: ${fe.message}`;
    return out;
}
/** @internal Formats a CTA block for human-readable TTY output. */
function formatHumanCta(cta) {
    const lines = ['', cta.description];
    for (const c of cta.commands) {
        const desc = c.description ? `  # ${c.description}` : '';
        lines.push(`  ${c.command}${desc}`);
    }
    return lines.join('\n');
}
/** @internal Type guard for sentinel results. */
function hasRequiredArgs(args) {
    return Object.values(args.shape).some((field) => field._zod.optout !== 'optional');
}
function isSentinel(value) {
    return typeof value === 'object' && value !== null && sentinel in value;
}
/** @internal Type guard for async generators returned by streaming `run` handlers. */
function isAsyncGenerator(value) {
    return (typeof value === 'object' &&
        value !== null &&
        Symbol.asyncIterator in value &&
        typeof value.next === 'function');
}
/** @internal Handles streaming output from an async generator `run` handler. */
async function handleStreaming(generator, ctx) {
    // Incremental: no explicit format (default toon), or explicit jsonl
    // Buffered: explicit json/yaml/toon/md
    const useJsonl = ctx.format === 'jsonl';
    const incremental = useJsonl || (!ctx.formatExplicit && ctx.format === 'toon');
    if (incremental) {
        // Incremental output: write each chunk as it arrives
        try {
            let returnValue;
            while (true) {
                const { value, done } = await generator.next();
                if (done) {
                    returnValue = value;
                    break;
                }
                if (isSentinel(value)) {
                    const tagged = value;
                    if (tagged[sentinel] === 'error') {
                        if (useJsonl)
                            ctx.writeln(JSON.stringify({
                                type: 'error',
                                ok: false,
                                error: {
                                    code: tagged.code,
                                    message: tagged.message,
                                    ...(tagged.retryable !== undefined
                                        ? { retryable: tagged.retryable }
                                        : undefined),
                                },
                            }));
                        else
                            ctx.writeln(formatHumanError({ code: tagged.code, message: tagged.message }));
                        ctx.exit(1);
                        return;
                    }
                }
                if (useJsonl)
                    ctx.writeln(JSON.stringify({ type: 'chunk', data: value }));
                else if (ctx.renderOutput)
                    ctx.writeln(Formatter.format(value, ctx.format));
            }
            // Handle return value — error() or ok() sentinel
            if (isSentinel(returnValue) && returnValue[sentinel] === 'error') {
                const err = returnValue;
                if (useJsonl)
                    ctx.writeln(JSON.stringify({
                        type: 'error',
                        ok: false,
                        error: {
                            code: err.code,
                            message: err.message,
                            ...(err.retryable !== undefined ? { retryable: err.retryable } : undefined),
                        },
                    }));
                else
                    ctx.writeln(formatHumanError({ code: err.code, message: err.message }));
                ctx.exit(1);
                return;
            }
            const cta = isSentinel(returnValue) && returnValue[sentinel] === 'ok'
                ? formatCtaBlock(ctx.name, returnValue.cta)
                : undefined;
            if (useJsonl)
                ctx.writeln(JSON.stringify({
                    type: 'done',
                    ok: true,
                    meta: {
                        command: ctx.path,
                        duration: `${Math.round(performance.now() - ctx.start)}ms`,
                        ...(cta ? { cta } : undefined),
                    },
                }));
            else if (cta)
                ctx.writeln(formatHumanCta(cta));
        }
        catch (error) {
            if (useJsonl)
                ctx.writeln(JSON.stringify({
                    type: 'error',
                    ok: false,
                    error: {
                        code: error instanceof IncurError ? error.code : 'UNKNOWN',
                        message: error instanceof Error ? error.message : String(error),
                    },
                }));
            else
                ctx.writeln(formatHumanError({
                    code: 'UNKNOWN',
                    message: error instanceof Error ? error.message : String(error),
                }));
            ctx.exit(1);
        }
    }
    else {
        // Buffered output: collect all chunks, write as single value
        const chunks = [];
        try {
            let returnValue;
            while (true) {
                const { value, done } = await generator.next();
                if (done) {
                    returnValue = value;
                    break;
                }
                if (isSentinel(value)) {
                    const tagged = value;
                    if (tagged[sentinel] === 'error') {
                        ctx.write({
                            ok: false,
                            error: {
                                code: tagged.code,
                                message: tagged.message,
                                ...(tagged.retryable !== undefined ? { retryable: tagged.retryable } : undefined),
                            },
                            meta: {
                                command: ctx.path,
                                duration: `${Math.round(performance.now() - ctx.start)}ms`,
                            },
                        });
                        ctx.exit(1);
                        return;
                    }
                }
                chunks.push(value);
            }
            if (isSentinel(returnValue) && returnValue[sentinel] === 'error') {
                const err = returnValue;
                ctx.write({
                    ok: false,
                    error: {
                        code: err.code,
                        message: err.message,
                        ...(err.retryable !== undefined ? { retryable: err.retryable } : undefined),
                    },
                    meta: {
                        command: ctx.path,
                        duration: `${Math.round(performance.now() - ctx.start)}ms`,
                    },
                });
                ctx.exit(1);
                return;
            }
            const cta = isSentinel(returnValue) && returnValue[sentinel] === 'ok'
                ? formatCtaBlock(ctx.name, returnValue.cta)
                : undefined;
            ctx.write({
                ok: true,
                data: chunks,
                meta: {
                    command: ctx.path,
                    duration: `${Math.round(performance.now() - ctx.start)}ms`,
                    ...(cta ? { cta } : undefined),
                },
            });
        }
        catch (error) {
            ctx.write({
                ok: false,
                error: {
                    code: error instanceof IncurError ? error.code : 'UNKNOWN',
                    message: error instanceof Error ? error.message : String(error),
                },
                meta: {
                    command: ctx.path,
                    duration: `${Math.round(performance.now() - ctx.start)}ms`,
                },
            });
            ctx.exit(1);
        }
    }
}
/** @internal Formats a CTA block into the output envelope shape. */
function formatCtaBlock(name, block) {
    if (!block || block.commands.length === 0)
        return undefined;
    return {
        description: block.description ?? 'Suggested commands:',
        commands: block.commands.map((c) => formatCta(name, c)),
    };
}
/** @internal Formats a CTA by prefixing the CLI name. Handles string and object forms. */
function formatCta(name, cta) {
    if (typeof cta === 'string')
        return { command: `${name} ${cta}` };
    const prefix = cta.command === name || cta.command.startsWith(`${name} `) ? '' : `${name} `;
    let cmd = `${prefix}${cta.command}`;
    if (cta.args)
        for (const [key, value] of Object.entries(cta.args))
            cmd += value === true ? ` <${key}>` : ` ${value}`;
    if (cta.options)
        for (const [key, value] of Object.entries(cta.options))
            cmd += value === true ? ` --${key} <${key}>` : ` --${key} ${value}`;
    return { command: cmd, ...(cta.description ? { description: cta.description } : undefined) };
}
/** @internal Builds the `--llms` manifest from the command tree. */
function buildManifest(commands, prefix = []) {
    return {
        version: 'incur.v1',
        commands: collectCommands(commands, prefix).sort((a, b) => a.name.localeCompare(b.name)),
    };
}
/** @internal Recursively collects leaf commands with their full paths. */
function collectCommands(commands, prefix) {
    const result = [];
    for (const [name, entry] of commands) {
        const path = [...prefix, name];
        if (isGroup(entry)) {
            result.push(...collectCommands(entry.commands, path));
        }
        else {
            const cmd = { name: path.join(' ') };
            if (entry.description)
                cmd.description = entry.description;
            const inputSchema = buildInputSchema(entry.args, entry.env, entry.options);
            const outputSchema = entry.output ? Schema.toJsonSchema(entry.output) : undefined;
            if (inputSchema || outputSchema) {
                cmd.schema = {};
                if (inputSchema?.args)
                    cmd.schema.args = inputSchema.args;
                if (inputSchema?.env)
                    cmd.schema.env = inputSchema.env;
                if (inputSchema?.options)
                    cmd.schema.options = inputSchema.options;
                if (outputSchema)
                    cmd.schema.output = outputSchema;
            }
            const examples = formatExamples(entry.examples);
            if (examples) {
                const cmdName = path.join(' ');
                cmd.examples = examples.map((e) => ({
                    ...e,
                    command: e.command ? `${cmdName} ${e.command}` : cmdName,
                }));
            }
            result.push(cmd);
        }
    }
    return result;
}
/** @internal Recursively collects leaf commands as `Skill.CommandInfo` for `--llms --format md`. */
function collectSkillCommands(commands, prefix, groups) {
    const result = [];
    for (const [name, entry] of commands) {
        const path = [...prefix, name];
        if (isGroup(entry)) {
            if (entry.description)
                groups.set(path.join(' '), entry.description);
            result.push(...collectSkillCommands(entry.commands, path, groups));
        }
        else {
            const cmd = { name: path.join(' ') };
            if (entry.description)
                cmd.description = entry.description;
            if (entry.args)
                cmd.args = entry.args;
            if (entry.env)
                cmd.env = entry.env;
            if (entry.hint)
                cmd.hint = entry.hint;
            if (entry.options)
                cmd.options = entry.options;
            if (entry.output)
                cmd.output = entry.output;
            const examples = formatExamples(entry.examples);
            if (examples) {
                const cmdName = path.join(' ');
                cmd.examples = examples.map((e) => ({
                    ...e,
                    command: e.command ? `${cmdName} ${e.command}` : cmdName,
                }));
            }
            result.push(cmd);
        }
    }
    return result.sort((a, b) => a.name.localeCompare(b.name));
}
/** @internal Formats examples into `{ command, description }` objects. `command` is the args/options suffix only. */
export function formatExamples(examples) {
    if (!examples || examples.length === 0)
        return undefined;
    return examples.map((ex) => {
        const parts = [];
        if (ex.args)
            for (const value of Object.values(ex.args))
                parts.push(String(value));
        if (ex.options)
            for (const [key, value] of Object.entries(ex.options))
                parts.push(`--${key} ${value}`);
        const result = { command: parts.join(' ') };
        if (ex.description)
            result.description = ex.description;
        return result;
    });
}
/** @internal Builds separate args, env, and options JSON Schemas. */
function buildInputSchema(args, env, options) {
    if (!args && !env && !options)
        return undefined;
    const result = {};
    if (args)
        result.args = Schema.toJsonSchema(args);
    if (env)
        result.env = Schema.toJsonSchema(env);
    if (options)
        result.options = Schema.toJsonSchema(options);
    return result;
}
/** @internal Scans argv for deprecated flags and writes warnings to stderr. */
function emitDeprecationWarnings(argv, optionsSchema, alias) {
    if (!optionsSchema)
        return;
    const shape = optionsSchema.shape;
    const deprecatedFlags = new Set();
    const deprecatedShorts = new Map();
    for (const key of Object.keys(shape)) {
        const meta = shape[key]?.meta?.();
        if (meta?.deprecated) {
            const kebab = key.replace(/[A-Z]/g, (c) => `-${c.toLowerCase()}`);
            deprecatedFlags.add(kebab);
            if (alias?.[key])
                deprecatedShorts.set(alias[key], kebab);
        }
    }
    if (deprecatedFlags.size === 0)
        return;
    for (const token of argv) {
        if (token.startsWith('--')) {
            const stripped = token.split('=')[0].slice(2);
            const raw = !deprecatedFlags.has(stripped) && stripped.startsWith('no-') ? stripped.slice(3) : stripped;
            if (deprecatedFlags.has(raw))
                process.stderr.write(`Warning: --${raw} is deprecated\n`);
        }
        else if (token.startsWith('-') && token.length >= 2) {
            for (const ch of token.slice(1))
                if (deprecatedShorts.has(ch))
                    process.stderr.write(`Warning: --${deprecatedShorts.get(ch)} is deprecated\n`);
        }
    }
}
