// incur@0.1.17/dist/Mcp.js
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import * as Schema from './Schema.js';
/** Starts a stdio MCP server that exposes commands as tools. */
export async function serve(name, version, commands, options = {}) {
    const server = new McpServer({ name, version });
    for (const tool of collectTools(commands, [])) {
        const mergedShape = {
            ...tool.command.args?.shape,
            ...tool.command.options?.shape,
        };
        const hasInput = Object.keys(mergedShape).length > 0;
        server.registerTool(tool.name, {
            ...(tool.description ? { description: tool.description } : undefined),
            ...(hasInput ? { inputSchema: mergedShape } : undefined),
        }, async (...callArgs) => {
            // registerTool passes (args, extra) when inputSchema is set, (extra) when not
            const params = hasInput ? callArgs[0] : {};
            const extra = hasInput ? callArgs[1] : callArgs[0];
            return callTool(tool, params, extra);
        });
    }
    const input = options.input ?? process.stdin;
    const output = options.output ?? process.stdout;
    const transport = new StdioServerTransport(input, output);
    await server.connect(transport);
}
/** @internal Executes a tool call and returns a CallToolResult. */
async function callTool(tool, params, extra) {
    try {
        const { args, options } = splitParams(params, tool.command);
        const parsedArgs = tool.command.args ? tool.command.args.parse(args) : {};
        const parsedOptions = tool.command.options ? tool.command.options.parse(options) : {};
        const parsedEnv = tool.command.env ? tool.command.env.parse(process.env) : {};
        const sentinel = Symbol.for('incur.sentinel');
        const okFn = (data) => ({ [sentinel]: 'ok', data });
        const errorFn = (opts) => ({ [sentinel]: 'error', ...opts });
        const raw = tool.command.run({
            args: parsedArgs,
            env: parsedEnv,
            options: parsedOptions,
            ok: okFn,
            error: errorFn,
        });
        // Streaming: send progress notifications per chunk, then return buffered result
        if (isAsyncGenerator(raw)) {
            const chunks = [];
            const progressToken = extra?._meta?.progressToken;
            let i = 0;
            for await (const chunk of raw) {
                if (typeof chunk === 'object' && chunk !== null && sentinel in chunk) {
                    const tagged = chunk;
                    if (tagged[sentinel] === 'error')
                        return {
                            content: [{ type: 'text', text: tagged.message ?? 'Command failed' }],
                            isError: true,
                        };
                }
                chunks.push(chunk);
                if (progressToken !== undefined && extra?.sendNotification)
                    await extra.sendNotification({
                        method: 'notifications/progress',
                        params: { progressToken, progress: ++i, message: JSON.stringify(chunk) },
                    });
            }
            return { content: [{ type: 'text', text: JSON.stringify(chunks) }] };
        }
        const awaited = await raw;
        if (typeof awaited === 'object' && awaited !== null && sentinel in awaited) {
            const tagged = awaited;
            if (tagged[sentinel] === 'error')
                return {
                    content: [{ type: 'text', text: tagged.message ?? 'Command failed' }],
                    isError: true,
                };
            return { content: [{ type: 'text', text: JSON.stringify(tagged.data ?? null) }] };
        }
        return { content: [{ type: 'text', text: JSON.stringify(awaited ?? null) }] };
    }
    catch (err) {
        return {
            content: [{ type: 'text', text: err instanceof Error ? err.message : String(err) }],
            isError: true,
        };
    }
}
/** @internal Type guard for async generators. */
function isAsyncGenerator(value) {
    return (typeof value === 'object' &&
        value !== null &&
        Symbol.asyncIterator in value &&
        typeof value.next === 'function');
}
/** @internal Recursively collects leaf commands as tool entries. */
function collectTools(commands, prefix) {
    const result = [];
    for (const [name, entry] of commands) {
        const path = [...prefix, name];
        if ('_group' in entry && entry._group)
            result.push(...collectTools(entry.commands, path));
        else {
            result.push({
                name: path.join('_'),
                description: entry.description,
                inputSchema: buildToolSchema(entry.args, entry.options),
                command: entry,
            });
        }
    }
    return result.sort((a, b) => a.name.localeCompare(b.name));
}
/** @internal Builds a merged JSON Schema from args and options Zod schemas. */
function buildToolSchema(args, options) {
    const properties = {};
    const required = [];
    for (const schema of [args, options]) {
        if (!schema)
            continue;
        const json = Schema.toJsonSchema(schema);
        Object.assign(properties, json.properties ?? {});
        required.push(...(json.required ?? []));
    }
    if (required.length > 0)
        return { type: 'object', properties, required };
    return { type: 'object', properties };
}
/** @internal Splits flat params into args vs options using schema shapes. */
function splitParams(params, command) {
    const argKeys = new Set(command.args ? Object.keys(command.args.shape) : []);
    const a = {};
    const o = {};
    for (const [key, value] of Object.entries(params)) {
        if (argKeys.has(key))
            a[key] = value;
        else
            o[key] = value;
    }
    return { args: a, options: o };
}
