// @effect/opentelemetry@0.61.0/dist/esm/Otlp.js
import * as HttpClientRequest from "@effect/platform/HttpClientRequest";
import { flow } from "effect/Function";
import * as Layer from "effect/Layer";
import * as OtlpLogger from "./OtlpLogger.js";
import * as OtlpMetrics from "./OtlpMetrics.js";
import * as OtlpSerialization from "./OtlpSerialization.js";
import * as OtlpTracer from "./OtlpTracer.js";
/**
 * Creates an OTLP layer.
 *
 * @since 1.0.0
 * @category Layers
 */
export const layer = options => {
  const baseReq = HttpClientRequest.get(options.baseUrl);
  const url = path => HttpClientRequest.appendUrl(baseReq, path).url;
  return Layer.mergeAll(OtlpLogger.layer({
    replaceLogger: options.replaceLogger,
    url: url("/v1/logs"),
    resource: options.resource,
    headers: options.headers,
    exportInterval: options.loggerExportInterval,
    maxBatchSize: options.maxBatchSize,
    shutdownTimeout: options.shutdownTimeout,
    excludeLogSpans: options.loggerExcludeLogSpans
  }), OtlpMetrics.layer({
    url: url("/v1/metrics"),
    resource: options.resource,
    headers: options.headers,
    exportInterval: options.metricsExportInterval,
    shutdownTimeout: options.shutdownTimeout
  }), OtlpTracer.layer({
    url: url("/v1/traces"),
    resource: options.resource,
    headers: options.headers,
    exportInterval: options.tracerExportInterval,
    maxBatchSize: options.maxBatchSize,
    context: options.tracerContext,
    shutdownTimeout: options.shutdownTimeout
  }));
};
/**
 * Creates an OTLP layer with JSON serialization.
 *
 * @since 1.0.0
 * @category Layers
 */
export const layerJson = /*#__PURE__*/flow(layer, /*#__PURE__*/Layer.provide(OtlpSerialization.layerJson));
/**
 * Creates an OTLP layer with Protobuf serialization.
 *
 * @since 1.0.0
 * @category Layers
 */
export const layerProtobuf = /*#__PURE__*/flow(layer, /*#__PURE__*/Layer.provide(OtlpSerialization.layerProtobuf));
