// @effect/opentelemetry@0.61.0/dist/cjs/OtlpSerialization.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.layerProtobuf = exports.layerJson = exports.OtlpSerialization = void 0;
var HttpBody = _interopRequireWildcard(require("@effect/platform/HttpBody"));
var Context = _interopRequireWildcard(require("effect/Context"));
var Layer = _interopRequireWildcard(require("effect/Layer"));
var OtlpProtobuf = _interopRequireWildcard(require("./internal/otlpProtobuf.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * OtlpSerialization service for tree-shakable protobuf support.
 *
 * This module provides the `OtlpSerialization` service that abstracts the
 * encoding of OTLP telemetry data to HttpBody.
 *
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category Tags
 */
class OtlpSerialization extends /*#__PURE__*/Context.Tag("@effect/opentelemetry/OtlpSerialization")() {}
/**
 * JSON serializer layer for OTLP telemetry data.
 *
 * It encodes telemetry data as JSON with `application/json` content type.
 *
 * @since 1.0.0
 * @category Layers
 */
exports.OtlpSerialization = OtlpSerialization;
const layerJson = exports.layerJson = /*#__PURE__*/Layer.succeed(OtlpSerialization, {
  traces: data => HttpBody.unsafeJson(data),
  metrics: data => HttpBody.unsafeJson(data),
  logs: data => HttpBody.unsafeJson(data)
});
/**
 * Protobuf serializer layer for OTLP telemetry data.
 *
 * This serializer encodes telemetry data using Protocol Buffers binary
 * format with `application/x-protobuf` content type. It provides more
 * efficient wire format compared to JSON.
 *
 * @since 1.0.0
 * @category Layers
 */
const layerProtobuf = exports.layerProtobuf = /*#__PURE__*/Layer.succeed(OtlpSerialization, {
  traces: data => HttpBody.uint8Array(OtlpProtobuf.encodeTracesData(data), "application/x-protobuf"),
  metrics: data => HttpBody.uint8Array(OtlpProtobuf.encodeMetricsData(data), "application/x-protobuf"),
  logs: data => HttpBody.uint8Array(OtlpProtobuf.encodeLogsData(data), "application/x-protobuf")
});
