// @effect/opentelemetry@0.61.0/dist/cjs/Otlp.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.layerProtobuf = exports.layerJson = exports.layer = void 0;
var HttpClientRequest = _interopRequireWildcard(require("@effect/platform/HttpClientRequest"));
var _Function = require("effect/Function");
var Layer = _interopRequireWildcard(require("effect/Layer"));
var OtlpLogger = _interopRequireWildcard(require("./OtlpLogger.js"));
var OtlpMetrics = _interopRequireWildcard(require("./OtlpMetrics.js"));
var OtlpSerialization = _interopRequireWildcard(require("./OtlpSerialization.js"));
var OtlpTracer = _interopRequireWildcard(require("./OtlpTracer.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * Creates an OTLP layer.
 *
 * @since 1.0.0
 * @category Layers
 */
const layer = options => {
  const baseReq = HttpClientRequest.get(options.baseUrl);
  const url = path => HttpClientRequest.appendUrl(baseReq, path).url;
  return Layer.mergeAll(OtlpLogger.layer({
    replaceLogger: options.replaceLogger,
    url: url("/v1/logs"),
    resource: options.resource,
    headers: options.headers,
    exportInterval: options.loggerExportInterval,
    maxBatchSize: options.maxBatchSize,
    shutdownTimeout: options.shutdownTimeout,
    excludeLogSpans: options.loggerExcludeLogSpans
  }), OtlpMetrics.layer({
    url: url("/v1/metrics"),
    resource: options.resource,
    headers: options.headers,
    exportInterval: options.metricsExportInterval,
    shutdownTimeout: options.shutdownTimeout
  }), OtlpTracer.layer({
    url: url("/v1/traces"),
    resource: options.resource,
    headers: options.headers,
    exportInterval: options.tracerExportInterval,
    maxBatchSize: options.maxBatchSize,
    context: options.tracerContext,
    shutdownTimeout: options.shutdownTimeout
  }));
};
/**
 * Creates an OTLP layer with JSON serialization.
 *
 * @since 1.0.0
 * @category Layers
 */
exports.layer = layer;
const layerJson = exports.layerJson = /*#__PURE__*/(0, _Function.flow)(layer, /*#__PURE__*/Layer.provide(OtlpSerialization.layerJson));
/**
 * Creates an OTLP layer with Protobuf serialization.
 *
 * @since 1.0.0
 * @category Layers
 */
const layerProtobuf = exports.layerProtobuf = /*#__PURE__*/(0, _Function.flow)(layer, /*#__PURE__*/Layer.provide(OtlpSerialization.layerProtobuf));
